### PR TITLE
[AEA-1222] add aea version compatibility in 'aea upgrade'

### LINKED ISSUE #2138: `aea upgrade` doesn't consider AEA version compatibility
**Is your feature request related to a problem? Please describe.**

The command `aea upgrade` doesn't consider AEA version compatibility when pulling packages from the registry.

E.g. suppose I am currently at version 0.7.0 with library and upgrading to 0.8.0 when lib is latest at 0.9.0. I would need to pull the latest packages for 0.8.0, and exclude the ones that use 0.9.0.

**Describe the solution you'd like**

Make `aea upgrade` to take into account the current version of the AEA library when looking for latest packages, i.e. comparing the `aea` version installed and `aea_version` field of the configurations.

**Describe alternatives you've considered**

n/a

**Additional context**

When the following issue on remote registry is addressed:
https://github.com/fetchai/agents-registry/issues/262

solving this issue is relatively easy in `--remote` mode.

### PR DESCRIPTION
**Pull request by marcofavorito**
## Proposed changes

Enable AEA version compatibility in `aea upgrade`, after https://github.com/fetchai/agents-registry/pull/275/
This means packages are searched among the one compatible with the `aea` version of the currently installed framework.

## Fixes

Fixes #2138 

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [ ] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [ ] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that code coverage does not decrease.
- [ ] I have checked that the documentation about the `aea cli` tool works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

n/a

### CODE REVIEW

File: aea/cli/upgrade.py [hunk 1]
```diff
@@ -301,7 +305,7 @@ def upgrade(self) -> bool:
                 f"Cannot remote path {current_path}. Error: {str(e)}."
             )
 
-        fetch_agent(self.ctx, agent_package_id.public_id, alias=self._TEMP_ALIAS)
+        fetch_agent(self.ctx, new_item, alias=self._TEMP_ALIAS)
```
Comment (marcofavorito) [inline aea/cli/upgrade.py:308]: this was wrong

File: tests/test_cli/test_upgrade.py [hunk 2]
```diff
@@ -1197,4 +1205,30 @@ def test_run(self, mock_click_echo, mock_get_latest_version):
...
+        "Test upgrade."
+        result = self.run_cli_command("upgrade", "--remote", "-y", cwd=self._get_cwd())
+        assert result.exit_code == 0
+
+        # compare with latest fetched agent.
+        ignore = [DEFAULT_AEA_CONFIG_FILE] + filecmp.DEFAULT_IGNORES
+        dircmp = filecmp.dircmp(self.current_agent_context, self.EXPECTED, ignore=ignore)
+        _left_only, _right_only, diff = dircmp_recursive(dircmp)
+        assert diff == set()
+"""
```
Comment (marcofavorito) [inline tests/test_cli/test_upgrade.py:1234]: this can't work atm (see below)

**Comment by marcofavorito** [comment]
The PR cannot work atm because the AEA registry filter on `aea_version` with `:latest` public id version doesn't give the expected result.

E.g. the following code:
```python
from aea.cli.registry.utils import request_api
request_api("GET", "/skills/fetchai/echo/latest", params=dict(aea_version="0.8.0"))
```
returns `fetchai/echo:0.13.0`  (the latest version, as if the aea_version constraint wasn't specified), and not `fetchai/echo:0.12.0` (the latest fetchai/echo skill compatible with aea==0.8.0).

https://aea-registry.fetch.ai/details/skill/fetchai/echo/0.13.0
https://aea-registry.fetch.ai/details/skill/fetchai/echo/0.12.0

I might miss something silly though.

**Review by DavidMinarsch — APPROVED** [review]
LGTM

**Pull request merged**